transports/websocket/src/framed.rs

```diff
@@ -191,9 +284,17 @@ where
 
                         let ws_key = {
                             let request = server.receive_request()
-                                .map_err(|e| Error::Handshake(Box::new(e)))
-                                .await?;
-                            request.into_key()
+                                .await;
+                            match request {
+                                Ok(request) => request.into_key(),
+                                Err(err) => return match err {
+                                    handshake::Error::Io(_) => Err(Error::Handshake(Box::new(err))),
+                                    _ => Ok(EitherConnection::Fallback(Fallback {
```

Comment:
One way or another we handle errors the wrong way here. By continuing with the connection unless there is an I/O error we may ignore actual handshake errors. To minimise the risk we should only fall back if we encounter a `handshake::Error::Http` (HTTP parsing failed), or a `handshake::Error::HeaderNotFound(name)` with `name` equal to (case insensitive) "Upgrade" (missing upgrade header), or a `handshake::Error::UnexpectedHeader(name)` with `name` equal to (case insensitive) "Upgrade" (not a websocket upgrade). The last two conditions cover cases where the underlying transport sends well formed HTTP. The approach is still not very good protocol design as malformed HTTP requests may still be intended to be handshake messages.

Reply:
👍